### PR TITLE
Add support for fixed-length arrays

### LINKED ISSUE #17: Support for fixed size arrays
I have a use case where I want to read in and write out binary data that consists of a struct that has multiple fields, where several of the fields are fixed size arrays of primitive types or dataclass-structs.

## Example
```c
struct small_struct_a {
    u8 field_1;
    u16 field_2;
};

struct big_struct {
    u16 field_1;
    u16 field_2[4];                 // Fixed-size array of 4 unsigned 16bit integers
    small_struct_a main_a_instance;
    small_struct_a all_as[10];      // Fixed size array of 10 sub-structs
};
```

## Questions
Does this library currently support the following features? If so, can you give an example of how to use them? (I looked through the readme, but only saw examples for fixed size byte arrays)

* Fixed-size arrays of primitives, ex. `field_2` of `big_struct`
* Fixed-size arrays of sub-structs of fixed size, ex. `all_as` of `big_struct`

If these features are not currently supported, is there any interest in or plans of supporting them in the future?

### PR DESCRIPTION
**Pull request by ExcaliburZero**
Closes #17

Adds support for fields that are fixed-length arrays.

### Simple
```python
from typing import Annotated

@dcs.dataclass_struct()
class FixedLength:
    fixed: Annotated[list[int], 5]
```

```python
>>> FixedLength.from_packed(FixedLength([1, 2, 3, 4, 5]).pack())
FixedLength(fixed=[1, 2, 3, 4, 5])
```

### Arrays of structs
```python
from typing import Annotated

@dcs.dataclass_struct()
class Vector2d:
    x: float
    y: float

@dcs.dataclass_struct()
class FixedLength:
    fixed: Annotated[list[Vector2d], 3]
```

```python
>>> FixedLength.from_packed(FixedLength([Vector2d(1.0, 2.0), Vector2d(3.0, 4.0), Vector2d(5.0, 6.0)]).pack())
FixedLength(fixed=[Vector2d(x=1.0, y=2.0), Vector2d(x=3.0, y=4.0), Vector2d(x=5.0, y=6.0)])
```

### Multi-dimensional arrays
```python
from typing import Annotated

@dcs.dataclass_struct()
class TwoDimArray:
    fixed: Annotated[list[Annotated[list[int], 2]], 3]
```

```python
>>> TwoDimArray.from_packed(TwoDimArray([[1, 2], [3, 4], [5, 6]]).pack())
TwoDimArray(fixed=[[1, 2], [3, 4], [5, 6]])
```

### typing.List instead of list
```python
from typing import Annotated, List

@dcs.dataclass_struct()
class FixedLength:
    fixed: Annotated[List[int], 5]
```

```python
>>> FixedLength.from_packed(FixedLength([1, 2, 3, 4, 5]).pack())
FixedLength(fixed=[1, 2, 3, 4, 5])
```

### CODE REVIEW

**Review by harrymander — CHANGES_REQUESTED** [review]
Thanks for working on this! There are just a few extra edge cases that need tests. I also noticed that no error is raised when using lists of nested structs with different byteorders/sizes. E.g. the following test fails:

```python
@pytest.mark.parametrize(
    "nested_size_byteorder,container_size_byteorder",
    itertools.combinations(ALL_VALID_SIZE_BYTEORDER_PAIRS, 2),
)
def test_list_of_dataclass_structs_with_mismatched_size_and_byteorder_fails(
    nested_size_byteorder, container_size_byteorder
) -> None:
    nested_size, nested_byteorder = nested_size_byteorder
    container_size, container_byteorder = container_size_byteorder
    exp_msg = f"byteorder and size of nested dataclass-struct does not \
match that of container (expected '{container_size}' size and \
'{container_byteorder}' byteorder, got '{nested_size}' size and \
'{nested_byteorder}' byteorder)"

    with pytest.raises(TypeError, match=f"^{escape(exp_msg)}$"):

        @dcs.dataclass_struct(size=nested_size, byteorder=nested_byteorder)
        class Nested:
            pass

        @dcs.dataclass_struct(
            size=container_size,
            byteorder=container_byteorder,
        )
        class _:
            y: Annotated[list[Nested], 2]
```

**Comment by harrymander** [comment]
I had a go at tidying up some of the annotation parsing logic and packing/unpacking of lists. See here: https://github.com/harrymander/dataclasses-struct/commits/refactor/. Wanted to avoid having to call  `_resolve_field` to do more parsing of type annotations during unpacking - what do  you think?

Also currently padding isn't handled in list fields, e.g. 

```python
padded_list: Annotated[Annotated[list[int], PadBefore(2)], 2]
```

File: dataclasses_struct/dataclass.py [hunk 1]
```diff
@@ -194,18 +241,55 @@ def format(self) -> str:
...
+                    "fixed-length array value type is not a dataclass struct "
+                    f"or a builtin type: {value_type}"
+                )
+
+            return type_.format() * self.n
+
+    def validate_default(self, val: list[T]) -> None:
+        if len(val) > self.n:
+            raise ValueError(
+                f"fixed-length array cannot be longer than {self.n} elements"
```
[the remark this answers is not in the document]
Comment (ExcaliburZero) [reply dataclasses_struct/dataclass.py]: Changed it to check for exact list length (not too short, not too long) and added a unit test for it in 73d4e979b2aa2805aeec5426f73d045b4cffcfae

File: dataclasses_struct/dataclass.py [hunk 2]
```diff
@@ -353,6 +474,17 @@ def _make_class(
...
 
+def _is_generic_alias(type_: type) -> bool:
+    """
+    Check if the given type is either kind of GenericAlias (public vs internal
+    to typing).
+
+    This is needed since types like list use GenericAlias, while
+    types like typing.List use _GenericAlias.
+    """
+    return type_ is GenericAlias or type_ is typing._GenericAlias  # type: ignore[attr-defined]
```
[the remark this answers is not in the document]
Comment (ExcaliburZero) [reply dataclasses_struct/dataclass.py]: Agreed. It's definitely better to not rely on having it.

It's now removed in your refactor in 8cd7a124adab409b4c8233d5ad2486387059ac13

File: dataclasses_struct/dataclass.py [hunk 3]
```diff
@@ -77,35 +80,79 @@ def __init__(
...
+                        value_type.__dataclass_struct__._init_from_args(args)
+                    )
+                elif get_origin(value_type) == Annotated:
+                    arg_array.extend(
+                        list(
+                            _DataclassStructInternal._generate_args_recursively(
+                                args, None, value_type
+                            )
+                        )
+                    )
```
[the remark this answers is not in the document]
Comment (ExcaliburZero) [reply dataclasses_struct/dataclass.py]: Fixed in your refactor in a2b39916f4a1f30ee556c160a675880b1dcb4311

File: test/test_pack_unpack.py [hunk 4]
```diff
@@ -191,6 +191,86 @@ class Container:
...
+    class T:
+        x: Annotated[List[int], 5]  # noqa: UP006
+
+    t = T([1, 2, 3, 4, 5])
+    packed = t.pack()
+    unpacked = T.from_packed(packed)
+    assert isinstance(unpacked.x, list)
+    assert t == unpacked
+
+
```
[the remark this answers is not in the document]
Comment (ExcaliburZero) [reply test/test_pack_unpack.py]: Added in c7a3ccdf73144881a9747fe245c2d2d04cc68cc3

File: test/test_pack_unpack.py [hunk 5]
```diff
@@ -191,6 +191,86 @@ class Container:
...
+        ]
+    )
+    packed = t.pack()
+    unpacked = T.from_packed(packed)
+    assert isinstance(unpacked.x, list)
+    assert t == unpacked
+
+
+@parametrize_all_sizes_and_byteorders()
+def test_pack_unpack_list(size, byteorder) -> None:
```
[the remark this answers is not in the document]
Comment (ExcaliburZero) [reply test/test_pack_unpack.py]: Changed in e6e0aaa311a209785940d8e2653416a054202b26

File: test/test_fields.py [hunk 6]
```diff
@@ -167,6 +180,32 @@ class _:
...
+@parametrize_all_sizes_and_byteorders()
+def test_invalid_array_length_fails(size, byteorder, length: int) -> None:
+    with pytest.raises(
+        ValueError,
+        match=r"^fixed-length array length must be positive non-zero int$",
+    ):
+
+        @dcs.dataclass_struct(size=size, byteorder=byteorder)
+        class _:
+            x: Annotated[list[int], length]
```
[the remark this answers is not in the document]
Comment (ExcaliburZero) [reply test/test_fields.py]: Changed in e6e0aaa311a209785940d8e2653416a054202b26

File: test/test_pack_unpack.py [hunk 7]
```diff
@@ -191,6 +191,86 @@ class Container:
     assert c == unpacked
 
 
+@parametrize_all_sizes_and_byteorders()
+def test_pack_unpack_array_of_primatives(size, byteorder) -> None:
```
[the remark this answers is not in the document]
Comment (ExcaliburZero) [reply test/test_pack_unpack.py]: Fixed in 22705eb7d466b9c01a850f8c1372d9a5e68f6f46
